sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py

```diff
@@ -84,92 +59,89 @@ def _load_window_of_centroids_and_features(
                 n_window_features_per_work_group
             )
 
-    return _initialize_window_of_centroids, _load_window_of_centroids_and_features
-
-
-def _make_accumulate_dot_products_kernel_func(
-    n_samples, n_features, window_n_features, window_n_centroids, with_X_l2_norm, dtype
-):
-    # TODO: this factorization is disappointing but I don't think numba.func can support
-    # more than this level of abstraction. I tried using a const bool to swtich
-    # X_l2_norm computation on and off coupled with explicitly passing the `signature`
-    # argument, but to no result (the JIT will fail).
-
-    zero = dtype(0.0)
-
     @dpex.func
-    def _get_X_value(sample_idx, window_feature_idx, first_feature_idx, X_t):
-        feature_idx = window_feature_idx + first_feature_idx
-        if (feature_idx < n_features) and (sample_idx < n_samples):
-            # performance for the line thereafter relies on L1 cache
-            return X_t[feature_idx, sample_idx]
-        else:
-            return zero
+    def _initialize_results(results):
+        # Initialize the partial pseudo inertia dot product for each
+        # of the window_n_centroids centroids in the window.
+        for i in range(window_n_centroids):
+            results[i] = zero
 
-    @dpex.func
-    # fmt: off
-    def _accumulate_one_feature(
-        X_value,              # PARAM
-        window_feature_idx,   # PARAM
-        centroids_window,     # IN
-        dot_products          # OUT
-    ):
-    # fmt: on
-        # For this given feature, loop on all centroids in the current
-        # window and accumulate the dot products
-        for window_centroid_idx in range(window_n_centroids):
-            centroid_value = centroids_window[window_feature_idx, window_centroid_idx]
-            dot_products[window_centroid_idx] += centroid_value * X_value
+    if not load_centroid_window_half_l2_norm:
+        return _initialize_results, _load_window_of_centroids_and_features
 
-    if with_X_l2_norm:
+    if load_centroid_window_half_l2_norm:
 
         @dpex.func
         # fmt: off
-        def _accumulate_dot_products(
-            sample_idx,          # PARAM
-            first_feature_idx,   # PARAM
-            X_t,                 # IN
-            centroids_window,    # IN
-            dot_products,        # OUT
+        def _initialize_window_of_centroids(
+            local_work_id,                  # PARAM
+            first_centroid_idx,             # PARAM
+            centroids_half_l2_norm,         # IN
+            centroids_window_half_l2_norm,  # OUT
```

Comment:
Similar comment about the naming of this variable:

```suggestion
            centroids_half_l2_norm_window,  # OUT
```